File: ios_app/iosApp/Multiplatform/Plugin/Channels/Router/RouterChannel.swift
//
//  RouterChannel.swift
//  iosApp
//
//  Created by Liber Da Silva on 14/08/2020.
//

import Flutter

open class RouterChannel : NSObject {
    
    private let routerChannel: FlutterMethodChannel
    private var isChannelReady: Bool = false
    private var onReadyHandler: () -> () = {}
    private var returnToPlatformHandler: (_ routerResult: RouterResult?) -> () = {
        (routerResult) -> () in
    }
        
    init(with messenger: FlutterBinaryMessenger) {
        routerChannel = FlutterMethodChannel(
            name: routerChannelId,
            binaryMessenger: messenger
        )
        
        super.init()
        
        routerChannel.setMethodCallHandler(onMethodCall)
    }

    func onMethodCall(_ methodCall: FlutterMethodCall, result: FlutterResult) {
        switch methodCall.method {
            case reportChannelIsReady:
                isChannelReady = true
                onReadyHandler()
                
            case returnToPlatform:
                var routerResult: RouterResult?
                
                if let args = methodCall.arguments as? [String: Any],
                   let routerResultString = args[self.routerResult] as? String,
                   let routerResultJSON = routerResultString.data(using: .utf8) {
                    routerResult = try! JSONDecoder().decode(RouterResult.self, from: routerResultJSON)
                }
                                
                returnToPlatformHandler(routerResult)
            
            default: result(FlutterMethodNotImplemented)
        }
    }
    
    public func onReady(_ handler:@escaping () -> ()) {
        self.onReadyHandler = handler
        if (isChannelReady) {
            self.onReadyHandler()
        }
    }

    public func navigateTo(_ routerArgs: RouterArgs) throws {
        let jsonData = try JSONEncoder().encode(routerArgs)
        let jsonString = String(data: jsonData, encoding: .utf8)
        
        routerChannel.invokeMethod(navigateTo, arguments: jsonString)
    }

    public func setReturnToPlatformHandler(_ handler:@escaping (_ routerResult: RouterResult?) -> ()) {
        self.returnToPlatformHandler = handler
    }

    public func destroy() {
        routerChannel.setMethodCallHandler(nil)
    }

    // MARK: Constants
    let routerChannelId = "com.rappipay_demo/router"

    let reportChannelIsReady = "reportChannelIsReady"
    let navigateTo = "navigateTo"
    let returnToPlatform = "returnToPlatform"

    let routerResult = "routerResult"
}

File: android_app/app/src/main/java/com/rappipay_demo/android_app/multiplatform/plugin/channels/router/RouterChannel.kt
package com.rappipay_demo.android_app.multiplatform.plugin.channels.router

import com.google.gson.Gson
import com.rappipay_demo.android_app.multiplatform.plugin.channels.router.models.RouterArgs
import com.rappipay_demo.android_app.multiplatform.plugin.channels.router.models.RouterResult
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class RouterChannel : MethodChannel.MethodCallHandler {

    private lateinit var routerChannel: MethodChannel
    private var isChannelReady = false
    private var onReadyHandler: () -> Unit = {}
    private var returnToPlatformHandler: (RouterResult?) -> Unit = {}

    fun createChannel(messenger: BinaryMessenger) {
        routerChannel = MethodChannel(messenger, ROUTER_CHANNEL_ID)
        routerChannel.setMethodCallHandler(this)
    }

    override fun onMethodCall(
        methodCall: MethodCall,
        result: MethodChannel.Result
    ) {
        when (methodCall.method) {
            REPORT_CHANNEL_IS_READY -> {
                isChannelReady = true
                onReadyHandler()
            }
            RETURN_TO_PLATFORM -> {
                val routerResultString = methodCall.argument<String>(ROUTER_RESULT)
                var routerResult = Gson().fromJson(routerResultString, RouterResult::class.java)
                returnToPlatformHandler(routerResult)
            }
            else -> result.notImplemented()
        }
    }

    fun onReady(handler: () -> Unit) {
        this.onReadyHandler = handler
        if (isChannelReady) {
            this.onReadyHandler()
        }
    }

    fun navigateTo(routerArgs: RouterArgs) = routerChannel?.invokeMethod(NAVIGATE_TO, Gson().toJson(routerArgs))

    fun setReturnToPlatformHandler(handler: (RouterResult?) -> Unit) {
        this.returnToPlatformHandler = handler
    }

    fun destroy() = routerChannel.setMethodCallHandler(null)

    companion object {
        const val ROUTER_CHANNEL_ID = "com.rappipay_demo/router"

        const val REPORT_CHANNEL_IS_READY = "reportChannelIsReady"
        const val NAVIGATE_TO = "navigateTo"
        const val RETURN_TO_PLATFORM = "returnToPlatform"

        const val ROUTER_RESULT = "routerResult"
    }
}

